Search Exclusion and Search Widgets

Malkierian opened the pull request:
This adds two new things: widget search exclusion, and search widgets. The first allows for preventing things like entire popout windows from showing in the search. The second allows the creation, use, and indexing of Widgets in popout windows and custom widgets.

The widget search adds the widgets to a separate map that is searched in addition to the widgets registered in the main menu structure, but are drawn using the same function, `Menu::MenuDrawItem`. These additional widgets are displayed in a second column, but because they're actual widgets, they behave in the search results like those in the main menu structure.

![image](https://github.com/user-attachments/assets/aaab9e0e-2fbb-41d5-b7fb-159560110ab9)

https://github.com/user-attachments/assets/ab9352e5-e01b-4cef-8ab4-b27735dbad56

(First click is on check tracker background color, second on overlay timer).

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4029972917.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4030031352.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4030047347.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Implement widget search exclusion.
Exclude all external windows from the search.

Malkierian committed:
Add second column to search results for non-widget results.

Malkierian committed:
Implement extra search index.
Add various controls, audio editor, and tracker options to the extra search index.
Fix capitalization on a few options.
Add click-to-view functionality to extra search column.

Malkierian committed:
Add extraSearchWidgets for searching through non-menu widgets.
Convert Controls extraSearches to widgets. Used to display in-window, as well as provide reference to extraSearchWidgets.

Malkierian commented:
Adding Do Not Merge because I'm doing it a different way and it's not done yet.

Malkierian committed:
Swap Audio search to menu widget search system.

Malkierian committed:
Merge branch 'develop-blair' of github.com:Malkierian/Shipwright into menu-upgrades

Malkierian committed:
Fix missing static on a ShipInit instance.

Malkierian committed:
Implement SearchWidgets in popout windows.
Remove remaining vestiges of custom search.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into menu-upgrades

Malkierian committed:
Add SearchWidget for randomizing audio on seed gen.

Pepe20129 commented on soh/soh/SohGui/MenuTypes.h:
```diff
@@ -41,8 +41,8 @@ typedef enum {
-    WIDGET_COLOR_24, // color picker without alpha
-    WIDGET_COLOR_32, // color picker with alpha
+    WIDGET_CVAR_COLOR_PICKER, // color picker without alpha
+    WIDGET_COLOR_PICKER,      // color picker with alpha
```
These don't seem like the best names.

Malkierian replied:
I think it's rather the comments I need to update. I saw no reason to differentiate in the widget types between alpha and no alpha, that's handled in the options.

Malkierian replied:
Removed them.

Pepe20129 commented on soh/soh/SohGui/Menu.cpp:
```diff
@@ -797,21 +862,19 @@ void Menu::DrawElement() {
+            ImGui::SetCursorPosX((ImGui::GetWindowWidth() - ImGui::CalcTextSize("No results found").x) / 2);
+            ImGui::SetCursorPosY(ImGui::GetCursorPosY() + 10.0f);
+            ImGui::TextColored(ImVec4(1.0f, 1.0f, 1.0f, 0.4f), "No results found");
+        }
```
What's this change for?

Malkierian replied:
This is part of moving and keeping the Clear Search button at the top of the child rather than at the bottom.

Pepe20129 reviewed (dismissed):
Some small things.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into menu-upgrades

Malkierian committed:
Remove comments on color picker entries in WidgetType.

Malkierian committed:
Put all search results in middle column for width management.
Change combobox alignment and label position in search results to unify them all to left and above for better organization.
Add "Search Results" text to header of results page for clarification.

Malkierian committed:
clang
